### PR TITLE
Add pub option for public keys

### PR DESCRIPTION
**Pull request by paweljakubas**
For credentials from public key we need to have public keys without chain code. 
https://github.com/input-output-hk/cardano-wallet/pull/2253

This PR adds a way to get public key (without chain code) if `--pub` is sneaked. Otherwise extended public key is returned.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,31 +40,42 @@ east student silly already breeze enact seat trade few way online skin grass hum
...
+</details>
+
+<details>
+  <summary>How to generate a public stake key (public key without chain code)</summary>
+
+```
+$ cat recovery-phrase.prv | cardano-address key from-recovery-phrase Shelley \
+| cardano-address key child 1852H/1815H/0H/2/0 \
+| cardano-address key public --pub
+xpub16apaenn9ut6s40lcw3l8v68xawlrlq20z2966uzcx8jmv2q9uy7qh83kg9
```
[the remark this answers is not in the document]
Comment (paweljakubas) [reply README.md:78]: or more precisely I would like `addr_vk` if we are dealing with payment/delegating address and `stake_vk` if we are dealing with stake address (reward account). If not possible we will need a way to have `--bech32WithHrp prefix`

File: command-line/lib/Options/Applicative/Public.hs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    , publicOpt
+    ) where
+
+import Prelude
+
+import Options.Applicative
+    ( Parser, flag, long )
+
+
+data PublicType = WithChainCode | WithoutChainCode
```
Comment (rvl) [inline command-line/lib/Options/Applicative/Public.hs:20]: Does a single type and parse function need its own module?

File: command-line/lib/Command/Key/Public.hs [hunk 3]
```diff
@@ -40,13 +43,19 @@ mod liftCmd = command "public" $
         <> progDesc "Get the public counterpart of a private key"
         <> footerDoc (Just $ string $ mconcat
             [ "The private key is read from stdin."
+            , "Extended public key is default. In order to get public key"
+            , ", ei., without chain code use '--pub'."
```
Comment (rvl) [inline command-line/lib/Command/Key/Public.hs:47]: Do you know which use case is more common?
I am thinking that public key only might be a better default. Not sure though.
Another alternative could be to have completely separate subcommands -- e.g. `public` and `xpub`, or something like that. Then you could tell users to run `xpub` if they need to derive child keys.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rvl) [reply README.md:78]: Perhaps use pub/xpub only in cardano-address, and show how to use the bech32 util to change HRP.

File: README.md [hunk 4]
```diff
@@ -40,31 +40,42 @@ east student silly already breeze enact seat trade few way online skin grass hum
   <summary>How to generate a root private key</summary>
 
 ```
-$ cardano-address recovery-phrase generate --size 15 > recovery-phrase.prv
-$ cat recovery-phrase.prv | cardano-address key from-recovery-phrase Shelley
-xprv1fzu4e8cecxshgzzxzh7557sd8tffqreeq2je7fgsm7f02mq849vdupw7qwgxc3qawyqev0l8ew0f4fkp8hvr8mskz4hz6e6ejzjlevcskcl6lqpr07u7552fsfgteztuclse7luh4cp493zdhkrjdss0250cdw8n
+$ cat recovery-phrase.txt | cardano-address key from-recovery-phrase Shelley
```
Comment (KtorZ) [inline README.md:43]: Why this change ? I think it was clearer before with both instructions visible in the same excerpt.

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (KtorZ) [reply README.md:78]: What do you think of flipping this around and, have this being the default behavior, but require an explicit `--with-chain-code` flag to return the public key with chain code.
 
We should avoid returning the chain code unless the user knows what he/she is doing with it. Changing the default, we would also need to change a bit the error message when failing to parse bytes into a public key + chain code (so that it hints users in the right direction, that a 32-byte chain code is missing).

File: command-line/lib/Command/Key/Public.hs [hunk 5]
```diff
@@ -40,13 +43,19 @@ mod liftCmd = command "public" $
...
             ])
   where
     parser = Public
         <$> encodingOpt [humanReadablePart|xpub|]
+        <*> publicOpt
```
Comment (rvl) [inline command-line/lib/Command/Key/Public.hs:52]: Sometthing like this will let you change the HRP...

```suggestion
  parser = do
    pub <- publicOpt
    Public <$> encodingOpt (if pub == WithChainCode then [humanReadablePart|xpub|] else [humanReadablePart|pub|])
        <*> pure pub
```

(clean it up though)

**Review by rvl — COMMENTED** [review]
Looks good so far - do you think pub or xpub should be the "default"? Or should there be separate subcommands for pub and xpub?

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rvl) [reply README.md:78]: I like that idea of requiring the user to specify what kind of key they need. Such as:
-  `(--with-chain-code | --without-chain-code )`.
- `cardano-address key xpub` /  `cardano-address key public`
- `( --extended-key | --key-only )`

File: command-line/lib/Command/Key/Public.hs (hunk 3 above)
Comment (KtorZ) [reply command-line/lib/Command/Key/Public.hs:47]: More than which one is most common, I'd go for the "safest" from a privacy standpoint here. Since sharing your chain code allows other to create new (public) child keys from your public key, you may inadvertently gives access to an entire part of your sub-tree by sharing your extended public key. 

Plus, cardano-cli and other cardano-tools mostly deals with non-extended ed25519 keys only.

File: command-line/lib/Options/Applicative/Public.hs (hunk 2 above)
Comment (paweljakubas) [reply command-line/lib/Options/Applicative/Public.hs:22]: I somewhat just complied with the pattern we have for other things. So although I am slightly in favour with your opinion on this this would be outlier and may surprise anyone NEW looking at the code....

File: command-line/lib/Command/Key/Public.hs (hunk 3 above)
Comment (paweljakubas) [reply command-line/lib/Command/Key/Public.hs]: introduced `--with-chain-code` and `--without-chain-code` here https://github.com/input-output-hk/cardano-addresses/pull/87/commits/c2e3c7801beb75a5541c86a1a3d8e09f06d17459

File: command-line/lib/Command/Key/Public.hs (hunk 5 above)
Comment (paweljakubas) [reply command-line/lib/Command/Key/Public.hs:52]: thanks for suggestion! in the end solved this like that https://github.com/input-output-hk/cardano-addresses/pull/87/commits/4ad1a31aa59c09476b4ae27c6465833cc28425e3

File: README.md (hunk 4 above)
Comment (paweljakubas) [reply README.md:43]: @KtorZ  fair point. I was thinking here in the following way. We use `prv`, `pub`, `xpub`, `pub` for keys...and for me mnemonic sequence is just text.... If that does not persuade you I will revert this in coming commit. @rvl do you have an opinion on that?

Comment (rvl) [reply README.md:43]: Yes and yes
1. It would be better if the `recovery-phrase generate` command were added back to provide context.
2. `.txt` is a good extension to describe the content of the file.

File: command-line/lib/Command/Key/Public.hs (hunk 5 above)
Comment (rvl) [reply command-line/lib/Command/Key/Public.hs:52]: if you use monadic rather than applicative then the ??? placeholder is not needed.

File: command-line/lib/Command/Key/Public.hs [hunk 6]
```diff
@@ -40,13 +43,25 @@ mod liftCmd = command "public" $
...
+        <$> encodingOpt [humanReadablePart|???|]
+        <*> publicOpt
 
 run :: Cmd -> IO ()
-run Public{encoding} = do
+run Public{encoding, chainCode} = do
     xprv <- hGetXPrv stdin
     let xpub = toXPub xprv
-    hPutBytes stdout (xpubToBytes xpub) encoding
+    (bytes, encoding') <- case chainCode of
```
Comment (rvl) [inline command-line/lib/Command/Key/Public.hs:58]: This line doesn't need to be an IO action. It could be a plain let or where clause.

I would be inclined to set up hrp in the CLI Parser, and use a function.

```haskell
let toBytes = case chainCode of
  WithChainCode -> xpubToBytes
  WithoutChainCode -> xpubPublicKey
hPutBytes stdout (toBytes xpub) encoding
```

File: command-line/lib/Options/Applicative/Public.hs (hunk 2 above)
Comment (rvl) [reply command-line/lib/Options/Applicative/Public.hs:22]: OK

File: command-line/lib/Command/Key/Public.hs [hunk 7]
```diff
@@ -24,14 +24,17 @@ import Options.Applicative.Encoding
...
 import System.IO
     ( stdin, stdout )
 import System.IO.Extra
     ( hGetXPrv, hPutBytes )
 
 
-newtype Cmd = Public
+data Cmd = Public
     { encoding :: Encoding
+    , chainCode :: PublicType
```
Comment (rvl) [inline command-line/lib/Command/Key/Public.hs:37]: Parser will look better if the order of constructor fields are swapped.

**Review by rvl — APPROVED** [review]
A couple style issues, but otherwise good.

**Review by KtorZ — APPROVED** [review]
